tuf_conformance/test_updater_delegation_graphs.py

```diff
@@ -0,0 +1,151 @@
+
+          targets
+*.doc, *md / \ release/*/*
+          A   B
+ release/x/* / \ release/y/*.zip
+            C   D
+
+Test that Updater successfully finds the target files metadata,
+traversing the delegations as expected.
+"""
+delegations_tree = DelegationsTestCase(
+    delegations=[
+        DelegationTester("targets", "A", paths=["*.doc", "*.md"]),
+        DelegationTester("targets", "B", paths=["releases/*/*"]),
+        DelegationTester("B", "C", paths=["releases/x/*"]),
+        DelegationTester("B", "D", paths=["releases/y/*.zip"]),
+    ],
+    target_files=[
+        TestTarget("targets", b"targetfile content", "targetfile"),
+        TestTarget("A", b"README by A", "README.md"),
+        TestTarget("C", b"x release by C", "releases/x/x_v1"),
+        TestTarget("D", b"y release by D", "releases/y/y_v1.zip"),
+        TestTarget("D", b"z release by D", "releases/z/z_v1.zip"),
+    ],
+)
+
+targets: DataSet = {
+    "no delegations": TargetTestCase("targetfile", True, []),
+    "targetpath matches wildcard": TargetTestCase("README.md", True, ["A"]),
+    "targetpath with separators x": TargetTestCase("releases/x/x_v1", True, ["B", "C"]),
+    "targetpath with separators y": TargetTestCase(
+        "releases/y/y_v1.zip", True, ["B", "D"]
+    ),
+    "targetpath is not delegated by all roles in the chain": TargetTestCase(
+        "releases/z/z_v1.zip", False, ["B"]
+    ),
+}
+
+
+targets_ids = targets.keys()
+targets_cases = targets.values()
+
+
+@pytest.mark.parametrize("target", targets_cases, ids=targets_ids)
+def test_targetfile_search(
+    client: ClientRunner, server: SimulatorServer, target: TargetTestCase
+) -> None:
+    exp_files = [*TOP_LEVEL_ROLE_NAMES, *target.visited_order]
+    exp_calls = [(role, 1) for role in target.visited_order]
+
+    init_data, repo = server.new_test(client.test_name)
+    assert client.init_client(init_data) == 0
+    init_repo(repo, delegations_tree)
+    exp_target = repo.artifacts[target.targetpath].target_file
+
+    # Call explicitly refresh to simplify the expected_calls list
+    client.refresh(init_data)
+    repo.metadata_statistics.clear()
+    client.download_target(init_data, target.targetpath)
+    if target.found:
+        assert client.download_target(init_data, target.targetpath) == 0
+        downloaded_target_path = client.get_downloaded_target_path()
+        downloaded_target_bytes = client.get_downloaded_target_bytes()[0]
+        local_target_file = TargetFile.from_data(
+            downloaded_target_path, downloaded_target_bytes
+        )
+        assert local_target_file.to_dict() == exp_target.to_dict()
+    else:
+        assert client.download_target(init_data, target.targetpath) == 1
+
+    # repo prepends and appends [('root', 2), ('timestamp', None)]
+    # so we compare equality from the 2nd call to the 3rd-last
+    # call.
+    assert repo.metadata_statistics[2:-2] == exp_calls
+    assert client._files_exist(exp_files)
```

Comment:
should rebase and use `trusted_roles()`

Reply:
could also assert about repo.artifact_statistics if that's not too complicated

Reply:
Removed the file check